pkg/registry/core/service/ipallocator/controller/repair.go

```diff
@@ -196,84 +195,95 @@ func (c *Repair) runOnce() error {
 		return fmt.Errorf("unable to refresh the service IP block: %v", err)
 	}
 
-	var rebuilt, secondaryRebuilt *ipallocator.Range
-	rebuilt, err = ipallocator.NewCIDRRange(c.network)
-	if err != nil {
-		return fmt.Errorf("unable to create CIDR range: %v", err)
-	}
-
-	if c.shouldWorkOnSecondary() {
-		secondaryRebuilt, err = ipallocator.NewCIDRRange(c.secondaryNetwork)
-	}
-
-	if err != nil {
-		return fmt.Errorf("unable to create CIDR range: %v", err)
+	getFamilyByIP := func(ip net.IP) v1.IPFamily {
+		if netutil.IsIPv6(ip) {
+			return v1.IPv6Protocol
+		}
+		return v1.IPv4Protocol
 	}
 
 	// Check every Service's ClusterIP, and rebuild the state as we think it should be.
 	for _, svc := range list.Items {
+		// (khenidak) double check that this is the best place
+		// for this.
+		// ensures that ClusterIPs are set according to ClusterIP
+		if svc.Spec.ClusterIPs == nil && len(svc.Spec.ClusterIP) > 0 {
+			svc.Spec.ClusterIPs = []string{svc.Spec.ClusterIP}
+		}
+
 		if !helper.IsServiceIPSet(&svc) {
 			// didn't need a cluster IP
 			continue
 		}
-		ip := net.ParseIP(svc.Spec.ClusterIP)
-		if ip == nil {
-			// cluster IP is corrupt
-			c.recorder.Eventf(&svc, v1.EventTypeWarning, "ClusterIPNotValid", "Cluster IP %s is not a valid IP; please recreate service", svc.Spec.ClusterIP)
-			runtime.HandleError(fmt.Errorf("the cluster IP %s for service %s/%s is not a valid IP; please recreate", svc.Spec.ClusterIP, svc.Name, svc.Namespace))
-			continue
-		}
 
-		// mark it as in-use
-		actualAlloc := c.selectAllocForIP(ip, rebuilt, secondaryRebuilt)
-		switch err := actualAlloc.Allocate(ip); err {
-		case nil:
-			actualStored := c.selectAllocForIP(ip, stored, secondaryStored)
-			if actualStored.Has(ip) {
-				// remove it from the old set, so we can find leaks
-				actualStored.Release(ip)
+		for idx, ip := range svc.Spec.ClusterIPs {
+			ip := net.ParseIP(ip)
+			if ip == nil {
+				// cluster IP is corrupt
+				c.recorder.Eventf(&svc, v1.EventTypeWarning, "ClusterIPNotValid", "Cluster IP %s is not a valid IP; please recreate service", ip)
+				runtime.HandleError(fmt.Errorf("the cluster IP %s for service %s/%s is not a valid IP; please recreate", ip, svc.Name, svc.Namespace))
+				continue
+			}
+
+			var family v1.IPFamily
+			if idx >= len(svc.Spec.IPFamilies) {
+				// (khenidak) the get decorator for some reason does not work within
```

Comment:
Is this still true?  We should crack that one, shouldn't we?

Reply:
As discussed - this should work correctly, so all the dup logic can be removed and cetralized in the read decorator